### ISSUE TITLE
Server resonds with incorrect error code when client sends empty list of certificates

### ISSUE THREAD

**Issue by marten-seemann** [cloudflare/tls-tris#141]
When client authentication is requested by setting `ClientAuth`, the server requests a certificate from the client, but it neither verifies if the client actually sent a certificate (for `RequireAnyClientCert` and `RequireAndVerifyClientCert`).

**Comment by henrydcase** [comment]
When client won't provide the certificate then it get's an alert ``21``
```
New, TLSv1.2, Cipher is ECDHE-RSA-AES128-GCM-SHA256
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES128-GCM-SHA256
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 2C6008B17A94EC4B7F779EA73CEB420551BAC7C8717A07E82F1541EFA879FD6BEABE8BBF8F2A49CAB0C648023D5B7279
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1539265850
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
    Extended master secret: no
```

**Comment by henrydcase** [comment]
I've to say, I disagree - in case of TLSv1.3 (I'm not sure yet about TLSv1.2). It may not be very clear, but in case client doesn't send certificate, the server will return with an error. Indeed, returned error is incorrect, but again - it doesn't seem to me that authentication will work.

In more details, assuming all TLS handshake messages are sent, but Certificate message from client doesn't actually contain any certificate. In case of client authentication on server side:
1. During handshake processing code hits line 13.go:361 
```
certs := getCertsFromEntries(certMsg.certificates)
```
2. ``len(certs) == 0``
3. ``serverHandshakeState.processCertsFromClient`` is called
4. Function returns with ``nil,nil``
```
828: 	if len(certs) == 0 {
829: 		return nil, nil
830:	}
```
4. ``verifyPeerHandshakeSignature`` is called with ``pubKey = nil``
5. ``pickSignatureAlgorithm`` is called with ``pubkey=nil``
6. Function returns:
```
return 0, 0, 0, errors.New("tls: peer doesn't support any common signature algorithms")
```

**Comment by henrydcase** [comment]
Yep, I'm aware that we will have this issue if #140 is merged but #142 is not. And this would be quite a bug. I would like to avoid having single commit which introduces security issue.

Nevertheless, #140 and #142 should be merged, this is valuable contribution. Could you squash both to one single commit?

**Comment by marten-seemann** [comment]
Makes sense. #142 it is.
